File: src/limiter/SlidingWindow.js
import BaseLimiter from "./BaseLimiter.js";
import { loadLuaScript } from "../lua/loadLuaScript.js";

class SlidingWindow extends BaseLimiter {

    constructor({ redis, window, max }) {
        super({ redis });

        this.validatePositiveInteger(window, "window");
        this.validatePositiveInteger(max, "max");

        this.window = window;
        this.max = max;

        this.scriptSha = null;
    }

    async getScriptSha() {
        if (!this.scriptSha) {
            this.scriptSha = await loadLuaScript(
                this.redis,
                "slidingWindow.lua"
            );
        }

        return this.scriptSha;
    }

    async consume(key) {

        const now = Date.now();
        const member = `${now}-${Math.random()}`;

        let sha = await this.getScriptSha();

        try {
            const [allowed, count, retryAfter] = await this.redis.evalSha(sha, {
                keys: [key],
                arguments: [
                    now.toString(),
                    this.window.toString(),
                    this.max.toString(),
                    member
                ]
            });

            if (allowed === 1) {
                return this.createResponse({
                    allowed: true,
                    limit: this.max,
                    remaining: this.max - count
                });
            }

            return this.createResponse({
                allowed: false,
                limit: this.max,
                remaining: 0,
                retryAfter
            });

        } catch (err) {
            if (err.message.includes("NOSCRIPT")) {
                this.scriptSha = null;

                sha = await this.getScriptSha();

                const [allowed, count, retryAfter] = await this.redis.evalSha(sha, {
                    keys: [key],
                    arguments: [
                        now.toString(),
                        this.window.toString(),
                        this.max.toString(),
                        member
                    ]
                });

                if (allowed === 1) {
                    return this.createResponse({
                        allowed: true,
                        limit: this.max,
                        remaining: this.max - count
                    });
                }

                return this.createResponse({
                    allowed: false,
                    limit: this.max,
                    remaining: 0,
                    retryAfter
                });
            }

            throw err;
        }
    }

}

export default SlidingWindow;

File: src/lua/slidingWindow.lua
local key = KEYS[1]

local now = tonumber(ARGV[1])
local window = tonumber(ARGV[2])
local max = tonumber(ARGV[3])
local member = ARGV[4]

redis.call("ZREMRANGEBYSCORE", key, 0, now - window)

local count = redis.call("ZCARD", key)

if count >= max then
    local oldest = redis.call("ZRANGE", key, 0, 0, "WITHSCORES")
    local retryAfter = math.ceil((tonumber(oldest[2]) + window - now) / 1000)

    return {
        0,
        count,
        retryAfter
    }
end

redis.call("ZADD", key, now, member)
redis.call("EXPIRE", key, math.ceil(window / 1000))

return {
    1,
    count + 1,
    0
}